akd/src/proto/mod.rs

```diff
@@ -0,0 +1,287 @@
+        })
+    }
+}
+
+impl From<&crate::proof_structs::SingleAppendOnlyProof<Hasher>> for audit::SingleEncodedProof {
+    fn from(input: &crate::proof_structs::SingleAppendOnlyProof<Hasher>) -> Self {
+        let mut result = Self::new();
+        let inserted = input
+            .inserted
+            .iter()
+            .map(|item| item.into())
+            .collect::<Vec<_>>();
+        let unchanged = input
+            .unchanged_nodes
+            .iter()
+            .map(|item| item.into())
+            .collect::<Vec<_>>();
+
+        result.set_inserted(RepeatedField::from_vec(inserted));
+        result.set_unchanged(RepeatedField::from_vec(unchanged));
+        result
+    }
+}
+
+impl TryFrom<audit::SingleEncodedProof> for crate::proof_structs::SingleAppendOnlyProof<Hasher> {
+    type Error = String;
+
+    fn try_from(input: audit::SingleEncodedProof) -> Result<Self, Self::Error> {
+        let mut inserted = vec![];
+        let mut unchanged = vec![];
+        for item in input.get_inserted() {
+            inserted.push(item.try_into()?);
+        }
+        for item in input.get_unchanged() {
+            unchanged.push(item.try_into()?);
+        }
+        Ok(crate::proof_structs::SingleAppendOnlyProof {
+            inserted,
+            unchanged_nodes: unchanged,
+        })
+    }
+}
+
+// ************************ Helper Functions ************************ //
+
+const NAME_SEPARATOR: char = '/';
+
+/// The constructed blobs with naming encoding the
+/// blob name = "EPOCH/PREVIOUS_ROOT_HASH/CURRENT_ROOT_HASH"
+pub struct AuditBlob {
+    /// The name of the blob, which can be decomposed into logical components (phash, chash, epoch)
+    pub name: String,
+    /// The binary data comprising the blob contents
+    pub data: Vec<u8>,
+}
+
+impl AuditBlob {
+    fn convert_akd_err(err: AkdError) -> String {
+        err.to_string()
+    }
+
+    /// Decompose the blob's name into the (previous_hash, current_hash, epoch) tuple
+    pub fn decompose_name(name: &str) -> Result<(Digest, Digest, u64), String> {
+        let parts = name.split(NAME_SEPARATOR).collect::<Vec<_>>();
+        if parts.len() < 3 {
+            return Err(String::from(
+                "Name is malformed, there are not enough components to reconstruct!",
+            ));
+        }
+        let epoch: u64 = parts[0]
+            .parse()
+            .map_err(|_| format!("Failed to parse {} into an unsigned integer", parts[0]))?;
+        let previous_epoch = crate::serialization::to_digest::<Hasher>(
+            &hex::decode(parts[1])
+                .map_err(|_| format!("Failed to parse {} as a hex string into bytes", parts[1]))?,
+        )
+        .map_err(Self::convert_akd_err)?;
+        let current_epoch = crate::serialization::to_digest::<Hasher>(
+            &hex::decode(parts[2])
+                .map_err(|_| format!("Failed to parse {} as a hex string into bytes", parts[2]))?,
+        )
```

Comment:
Any for `crate::serialization::to_digest::<Hasher>` (see: hash_from_bytes)? Any reason we can't use that here?